src/libcore/iter/adapters/chain.rs

```diff
@@ -57,128 +53,99 @@ where
 
     #[inline]
     fn next(&mut self) -> Option<A::Item> {
-        match self.state {
-            ChainState::Both => match self.a.next() {
-                elt @ Some(..) => elt,
-                None => {
-                    self.state = ChainState::Back;
-                    self.b.next()
-                }
-            },
-            ChainState::Front => self.a.next(),
-            ChainState::Back => self.b.next(),
+        match fuse!(self.a.next()) {
+            None => fuse!(self.b.next()),
+            item => item,
         }
     }
 
     #[inline]
     #[rustc_inherit_overflow_checks]
     fn count(self) -> usize {
-        match self.state {
-            ChainState::Both => self.a.count() + self.b.count(),
-            ChainState::Front => self.a.count(),
-            ChainState::Back => self.b.count(),
+        match self {
+            Chain { a: Some(a), b: Some(b) } => a.count() + b.count(),
+            Chain { a: Some(a), b: None } => a.count(),
+            Chain { a: None, b: Some(b) } => b.count(),
```

Comment:
This may emit the code for count 4 times, instead of just 2, so consider doing this instead as `(match self.a { ... }) + (match self.b { ... })`.